compiler/qsc_rca/src/lib.rs

```diff
@@ -0,0 +1,841 @@
+pub struct CallableComputeProperties {
+    /// The application generator set for the callable's body.
+    pub body: ApplicationGeneratorSet,
+    /// The application generator set for the callable's adjoint specialization.
+    pub adj: Option<ApplicationGeneratorSet>,
+    /// The application generator set for the callable's controlled specialization.
+    pub ctl: Option<ApplicationGeneratorSet>,
+    /// The application generator set for the callable's controlled adjoint specialization.
+    pub ctl_adj: Option<ApplicationGeneratorSet>,
+}
+
+impl Display for CallableComputeProperties {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let mut indent = set_indentation(indented(f), 0);
+        write!(indent, "CallableComputeProperties:",)?;
+        indent = set_indentation(indent, 1);
+        write!(indent, "\nbody: {}", self.body)?;
+        match &self.adj {
+            Some(spec) => write!(indent, "\nadj: {spec}")?,
+            None => write!(indent, "\nadj: <none>")?,
+        }
+        match &self.ctl {
+            Some(spec) => write!(indent, "\nctl: {spec}")?,
+            None => write!(indent, "\nctl: <none>")?,
+        }
+        match &self.ctl_adj {
+            Some(spec) => write!(indent, "\nctl-adj: {spec}")?,
+            None => write!(indent, "\nctl-adj: <none>")?,
+        }
+        Ok(())
+    }
+}
+
+/// A set of compute properties associated to a callable or one of its elements, from which the properties of any
+/// particular call application can be derived.
+#[derive(Clone, Debug)]
+pub struct ApplicationGeneratorSet {
+    /// The inherent compute kind of a program element, which is determined by binding all the parameters it depends on
+    /// to static values.
+    pub inherent: ComputeKind,
+    /// Each element in the vector represents the compute kind(s) of a call application when the parameter associated to
+    /// the vector index is bound to a dynamic value.
+    pub(crate) dynamic_param_applications: Vec<ParamApplication>,
+}
+
+impl Display for ApplicationGeneratorSet {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let mut indent = set_indentation(indented(f), 0);
+        write!(indent, "ApplicationsGeneratorSet:",)?;
+        indent = set_indentation(indent, 1);
+        write!(indent, "\ninherent: {}", self.inherent)?;
+        write!(indent, "\ndynamic_param_applications:")?;
+        if self.dynamic_param_applications.is_empty() {
+            write!(indent, " <empty>")?;
+        } else {
+            indent = set_indentation(indent, 2);
+            for (param_index, param_application) in
+                self.dynamic_param_applications.iter().enumerate()
+            {
+                write!(indent, "\n[{param_index}]: {param_application}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ApplicationGeneratorSet {
+    #[must_use]
+    pub fn derive_application_compute_kind(&self, args_value_kinds: &[ValueKind]) -> ComputeKind {
```

Comment:
on the topic of naming, I think it would be helpful to have this reflect whatever verb gets used in the struct name or go with a simple `get`. So right now, I would lean toward either `generate_application_compute_kind` or `get_application_compute_kind`.

Reply:
Renamed.